File: src/main/kotlin/uz/rms/modules/v1/purchase/model/Purchase.kt
package uz.rms.modules.v1.purchase.model

import jakarta.persistence.*
import uz.rms.common.BaseEntity
import uz.rms.modules.v1.supplier.model.Supplier
import uz.rms.modules.v1.warehouse.model.Warehouse
import java.math.BigDecimal
import java.time.LocalDateTime

enum class PurchaseStatus {
    PAID,
    DRAFT,
    PENDING,
    ORDERED,
    RECEIVED,
    PARTIALLY_RECEIVED,
    CANCELLED,
    RETURNED
}

@Entity
@Table(name = "purchases")
class Purchase : BaseEntity() {
    @Column(nullable = false, unique = true)
    lateinit var invoiceNumber: String

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "supplier_id", nullable = false)
    lateinit var supplier: Supplier

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "warehouse_id", nullable = false)
    lateinit var warehouse: Warehouse

    @Column(nullable = false)
    var purchaseDate: LocalDateTime = LocalDateTime.now()

    @Column(columnDefinition = "TEXT")
    var notes: String? = null

    @Column(nullable = false, precision = 12, scale = 2)
    var subTotal: BigDecimal = BigDecimal.ZERO

    @Column(nullable = false, precision = 12, scale = 2)
    var taxAmount: BigDecimal = BigDecimal.ZERO

    @Column(nullable = false, precision = 12, scale = 2)
    var discountAmount: BigDecimal = BigDecimal.ZERO

    @Column(nullable = false, precision = 12, scale = 2)
    var totalAmount: BigDecimal = BigDecimal.ZERO

    @Column(nullable = false)
    var paidAmount: BigDecimal = BigDecimal.ZERO

    @Column(nullable = false)
    var dueAmount: BigDecimal = BigDecimal.ZERO

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    var status: PurchaseStatus = PurchaseStatus.PENDING

    @Column(nullable = false)
    var isPaid: Boolean = false

    @Column(name = "is_received")
    var isReceived: Boolean = false

    @Column(name = "payment_due_date")
    var paymentDueDate: LocalDateTime? = null

}

File: frontend/src/types/purchase.ts
import { BaseEntity } from './base';

export enum PurchaseStatus {
  DRAFT = 'DRAFT',
  PENDING = 'PENDING',
  ORDERED = 'ORDERED',
  RECEIVED = 'RECEIVED',
  PARTIALLY_RECEIVED = 'PARTIALLY_RECEIVED',
  CANCELLED = 'CANCELLED',
  RETURNED = 'RETURNED',
  PAID = 'PAID',
}

export interface PurchaseItem {
  id?: number;
  productId: number;
  productName: string;
  productCode: string;
  quantity: number;
  unitPrice: number;
  taxRate: number;
  taxAmount: number;
  discount: number;
  total: number;
  receivedQuantity?: number;
}

export interface Purchase extends BaseEntity {
  id: number;
  invoiceNumber: string;
  supplierId: number;
  supplierName: string;
  warehouseId: number;
  warehouseName: string;
  purchaseDate: string;
  notes?: string;
  subTotal: number;
  taxAmount: number;
  discountAmount: number;
  totalAmount: number;
  paidAmount: number;
  dueAmount: number;
  status: PurchaseStatus;
  isPaid: boolean;
  isReceived: boolean;
  paymentDueDate?: string;
  items: PurchaseItem[];
  payments?: Payment[];
}

export interface Payment {
  id: number;
  amount: number;
  paymentDate: string;
  paymentMethod: string;
  reference: string;
  notes?: string;
  createdBy: string;
  createdAt: string;
}

export interface PurchaseFilterParams {
  page?: number;
  size?: number;
  status?: PurchaseStatus;
  supplierId?: number;
  startDate?: string;
  endDate?: string;
  sortBy?: string;
  sortOrder?: 'asc' | 'desc';
}

export interface CreatePurchaseRequest {
  invoiceNumber: string;
  supplierId: number;
  warehouseId: number;
  purchaseDate: string;
  notes?: string;
  subTotal: number;
  taxAmount: number;
  discountAmount: number;
  totalAmount: number;
  paidAmount: number;
  paymentDueDate?: string;
  items: Array<{
    productId: number;
    quantity: number;
    unitPrice: number;
    taxRate: number;
    discount: number;
    total: number;
  }>;
}

export interface UpdatePurchaseRequest {
  invoiceNumber?: string;
  supplierId?: number;
  warehouseId?: number;
  purchaseDate?: string;
  notes?: string;
  subTotal?: number;
  taxAmount?: number;
  discountAmount?: number;
  totalAmount?: number;
  paidAmount?: number;
  paymentDueDate?: string;
  status?: string;
}
